components/global/core/io.cellery.observability.k8s.client/src/main/java/io/cellery/observability/k8s/client/ComponentsEventSource.java

```diff
@@ -0,0 +1,259 @@
+        }
+        // Register the custom resource kind cell to Kubernetes deserializer to perform deserialization of cell objects.
+        KubernetesDeserializer.registerCustomKind(CELL_CRD_GROUP + VERSION, CELL, Cell.class);
+
+        // Create client for cell resource
+        NonNamespaceOperation<Cell, CellList, DoneableCell, Resource<Cell, DoneableCell>> cellClient =
+                k8sClient.customResources(cellCRD, Cell.class, CellList.class, DoneableCell.class)
+                        .inNamespace(DEFAULT_NAMESPACE);
+
+        // Cell watcher for Cellery cell updates
+        cellClient.watch(new Watcher<Cell>() {
+            @Override
+            public void eventReceived(Action action, Cell cell) {
+                List<HTTP> httpObjectsList = cell.getSpec().getGatewayTemplate().getSpec().getHttp();
+                List<TCP> tcpObjectsList = cell.getSpec().getGatewayTemplate().getSpec().getTcp();
+                IngressTypes ingressList = new IngressTypes();
+                try {
+                    Map<String, Object> attributes = new HashMap<>();
+                    attributes.put(Constants.ATTRIBUTE_CELL, cell.getMetadata().getName());
+                    attributes.put(Constants.ATTRIBUTE_CREATION_TIMESTAMP,
+                            cell.getMetadata().getCreationTimestamp() == null
+                                    ? -1
+                                    : new SimpleDateFormat(Constants.K8S_DATE_FORMAT, Locale.US).parse(
+                                    cell.getMetadata().getCreationTimestamp()).getTime());
+                    attributes.put(Constants.ATTRIBUTE_ACTION, action.toString());
+                    attributes.put(LAST_KNOWN_ACTIVE_TIMESTAMP, 0L);
+
+                    // Checks and add HTTP as Ingress type
+                    addHTTPTypes(httpObjectsList, sourceEventListener, attributes, ingressList);
+                    // Checks and add TCP as Ingress type
+                    addTCPTypes(tcpObjectsList, sourceEventListener, attributes, ingressList);
+                    // Checks and add HTTP as Ingress type
+                    addWEBTypes(httpObjectsList, sourceEventListener, attributes, cell, ingressList);
+
+                } catch (ParseException e) {
+                    logger.error("Ignored cell change due to creation timestamp parse failure", e);
+                }
+            }
+
+            @Override
+            public void onClose(KubernetesClientException cause) {
+                if (cause == null) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("Kubernetes cell watcher closed successfully");
+                    }
+                } else {
+                    logger.error("Kubernetes cell watcher closed with error", cause);
+                }
+            }
+        });
+        if (logger.isDebugEnabled()) {
+            logger.debug("Created cell watcher");
+        }
+    }
+
+    @Override
+    public void disconnect() {
```

Comment:
IINM, you need to close the watch here as well.

Reply:
Noted. I will update it.